### PR TITLE
wrap doRequest calls for error handling

### PR DESCRIPTION
**Pull request by FFMMM**
### Overview

This PR builds on https://github.com/hashicorp/consul/pull/11054 to wrap all remaining `doRequest()` calls around a `require*()` call.

### Issue Related
https://github.com/hashicorp/consul/issues/10865

### Notes:
* mostly refactoring, no logic changes

### PR Checklist
* [x] `go fmt`
* [x] `go mod`
* [x] n/a pr/changelog -- no "public" api changes

### CODE REVIEW

File: api/agent.go
Comment (freddygv) [inline api/agent.go:644]: Why's that?

**Review by freddygv — COMMENTED** [review]
Thanks for following up here! I left some questions below

File: api/agent.go
Comment (kisunji) [inline api/agent.go:1245]: minor: remove newline here

File: api/agent.go
Comment (kisunji) [inline api/agent.go:1247]: I'm a little confused about the order in which `requireOK` is called here - should this be handled close to L1238 right after `doRequest(r)`?

File: api/agent.go
Comment (FFMMM) [reply api/agent.go:604]: Good point! There's two ways I understand this comment:

1. Should we return a `StatusError` from the `require` calls and handle them inside the body of the func?

2. Should we return a `StatusError` our of this func?

Comment (FFMMM) [reply api/agent.go:604]: Let's chat about 1) first.

From a cursory look, I was worried that we'd change the return format.

So I was thinking that some of callees of these two API paths (ie `/v1/agent/health/service/id/%v`) were error handling for the old err messages. 

In order to keep the same format, we'd have to wrap it the same so it'd end up looking something like

```
return HealthCritical, out, fmt.Errorf("Unexpected Error Code %v for %s", err.StatusCode, path)
```
--
That's why I also said that there isn't a lot of value to wrapping the call since there is sufficient "behavior" handling and a different error message returned.

Comment (FFMMM) [reply api/agent.go:604]: For 2), I think we could get away with sending out an `*api.StatusError`. It's just a matter of whether it makes sense to expose that. I've not looked too hard into what changes would be required to do so if any at all.

What do you think ?

File: api/agent.go
Comment (FFMMM) [reply api/agent.go:644]: tried to answer and consolidate the convo here --> https://github.com/hashicorp/consul/pull/11158/files#r717131243

File: api/agent.go
Comment (FFMMM) [reply api/agent.go:1245]: hm, I can certainly do that but I was hoping that any style nits were taken care of by `gofmt`

File: api/kv.go
Comment (FFMMM) [reply api/kv.go:132]: It does! I thought I had cleaned up the deadcode but I didn't it seems!

Let me rework that.

File: api/agent.go
Comment (FFMMM) [reply api/agent.go:1247]: Yes I think so -- thanks for spotting that!

PS: I think when I chose not to I forgot we actually check the err from `doRequest()` inside the require calls!

https://github.com/hashicorp/consul/blob/5c37819d0957f54242d46650e74756346e4e1b2c/api/api.go#L1107-L1111

Comment (FFMMM) [reply api/agent.go]: Let me follow up here to explain the old odd/ smelly ordering!

So in the snippet above, if there is an error during the `doRequest()`, and we swallow the `resp` object and throw it back as `nil`. 

So there was not an easy way to combine the first err check, where we want to return 0 for some reason to the callee of `udpateTokenOnce()`, with the synthetic error checks.

As of now, I tried to re-write the snippet of code to make more sense -- hopefully it's clearer now but please let me know if it's not.

File: api/agent.go
Comment (kisunji) [reply api/agent.go:605]: We should consistently return `api.StatusError`, especially in cases where error strings look similar (otherwise it might mislead clients to believe they're dealing with the custom type)

**Comment by dnephin** [comment]
I haven't had a chance to catch up on this discussion, but I think that suggested refactor makes sense at a high level. Assuming the common case is `requireOk` we could even add a `doRequestRequireOk` that contains most of those lines. For the few places that don't call `requireOk` we can keep them as separate calls.

**Comment by FFMMM** [comment]
hey @kisunji and @dnephin thank y'all for chiming in! let me agree with y'all below with some context and proposed  path:

> I think there's some refactoring that may need to be done for requireOK + requireHttpStatusCodes.

Yes that's actually something that @markan and I were chatting about when looking at the `api/agent.go` code. It's weird ergonomics that the `require()` funcs close the response body for `doRequest` errors and not for other errors.

I think I'd love to have a chance to address this refactor in a subsequent PR.

> Assuming the common case is requireOk we could even add a doRequestAndRequireOk that contains most of those lines. 

There's actually >130 usages of `requireOK()` calls when I looked at that yesterday. I'd argue that a refactor that does `doRequestAndRequireOk` would be a big diff and even one that touches the internals of `requireOK `to modify behavior would warrant its own separate PR. 

---

I'm hoping that we can merge this PR if it makes sense and I can take on the work to have a stab at refactoring the code in a subsequent PR. What do y'all think? 👍🏼 or 👎🏼  ?

**Review by kisunji — APPROVED** [review]
Sounds reasonable to defer the refactor to another PR!

low-prio todos from my perspective:
* revisit error-handling of `requireOK` to be more clear/explicit
* use `api.StatusError` in all `fmt.Errorf`'s where we embed the status code in the error message

File: api/agent.go
Comment (FFMMM) [reply api/agent.go:605]: > I was mostly asking about 2), since the error string looks similar to the one from StatusError

> We should consistently return api.StatusError, especially in cases where error strings look similar (otherwise it might mislead clients to believe they're dealing with the custom type)

+1; should be something that I can address per our chat below in a separate PR

merging these changes for now.

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/461161.

**Pull request merged**
